File: shoring/flywheel/src/main/java/com/panopset/flywheel/ListAudit.java
package com.panopset.flywheel;

import java.io.StringWriter;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;
import java.util.TreeMap;
import java.util.TreeSet;
import com.panopset.compat.Stringop;

public class ListAudit {
  
  Map<String, List<String>> map = Collections.synchronizedSortedMap(new TreeMap<> ());
  Set<String> names = Collections.synchronizedSortedSet(new TreeSet<>());

  public void add(String name, List<String> lines) {
    if (!names.contains(name)) {
      names.add(name);
    }
    for (String line : lines) {
      List<String> listNames = map.get(line);
      if (listNames == null) {
        listNames = new ArrayList<>();
        listNames.add(name);
        map.put(line, listNames);
      } else {
        if (!listNames.contains(name)) {
          listNames.add(name);
        }
      }
    }
  }

  public List<String> getReport() {
    List<String> rtn = new ArrayList<>();
    rtn.add(getHeaderLine());
    for (Entry<String, List<String>> e : map.entrySet()) {
      StringWriter sw = new StringWriter();
      sw.append(e.getKey());
      for (String name: names) {
        sw.append(",");
        if (e.getValue().contains(name)) {
          sw.append("*");
        }
      }
      rtn.add(sw.toString());
    }
    return rtn;
  }

  public String getReportText() {
    return Stringop.listToString(getReport());
  }
  
  private String getHeaderLine() {
    StringWriter sw = new StringWriter();
    for (String name : names) {
      sw.append(String.format(",%s", name));
    }
    return sw.toString();
  }
}

File: code/shoring/flywheel/src/main/kotlin/com/panopset/flywheel/ListAudit.kt
package com.panopset.flywheel

import com.panopset.compat.Stringop.listToString
import java.io.StringWriter
import java.util.*

class ListAudit {
    var map: MutableMap<String, MutableList<String>> = Collections.synchronizedSortedMap(TreeMap())
    var names: MutableSet<String> = Collections.synchronizedSortedSet(TreeSet())
    fun add(name: String, lines: List<String>) {
        if (!names.contains(name)) {
            names.add(name)
        }
        for (line in lines) {
            var listNames = map[line]
            if (listNames == null) {
                listNames = ArrayList()
                listNames.add(name)
                map[line] = listNames
            } else {
                if (!listNames.contains(name)) {
                    listNames.add(name)
                }
            }
        }
    }

    val report: List<String?>
        get() {
            val rtn: MutableList<String?> = ArrayList()
            rtn.add(headerLine)
            for ((key, value) in map) {
                val sw = StringWriter()
                sw.append(key)
                for (name in names) {
                    sw.append(",")
                    if (value.contains(name)) {
                        sw.append("*")
                    }
                }
                rtn.add(sw.toString())
            }
            return rtn
        }
    val reportText: String
        get() = listToString(report)
    private val headerLine: String
        get() {
            val sw = StringWriter()
            for (name in names) {
                sw.append(String.format(",%s", name))
            }
            return sw.toString()
        }
}
